demo/project_demo00-SecOps/simulator/src/main.rs

```diff
@@ -94,32 +99,92 @@ impl KafkaProducer {
             .set_log_level(RDKafkaLogLevel::Debug);
         let producer = ThreadedProducer::from_config(&producer_config).unwrap();
 
-        Self { producer }
+        Self { producer: Arc::new(producer) }
     }
 
-    pub fn send_to_topic<T>(&self, topic: &str, data: &[Vec<T>])
+    pub fn insert_into_topic<T, I>(&self, topic: &str, data: I, message_size: usize)
     where
         T: Serialize + Clone,
+        I: IntoIterator<Item = T>,
     {
-        for batch in data {
-            let mut writer = CsvWriterBuilder::new()
-                .has_headers(false)
-                .from_writer(Vec::with_capacity(batch.len() * 32));
+        self.send_to_topic(topic, data.into_iter().map(|x| (x, true)), message_size)
+    }
 
-            for val in batch.iter().cloned() {
-                writer.serialize(val).unwrap();
+    pub fn send_to_topic<T, I>(&self, topic: &str, data: I, message_size: usize)
```

Comment:
The naming is weird here. The name `message_size` implies (to my mind) that we're measuring the size of a message, but it's used as a maximum count. I'd either name it something like `max_updates` or change the send condition to something like `bytes.len() >= message_size`. (If you do the latter you don't need the `updates` variable anymore.)